java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -0,0 +1,228 @@
+        writableDatabase.useTransaction {
+            execSQL("DELETE FROM collection_entries")
+            execSQL("DELETE FROM collections")
+            execSQL("DELETE FROM entities")
+            execSQL("DELETE FROM field_values")
+            execSQL("DELETE FROM fields")
+            execSQL("DELETE FROM number_primitive_values")
+            execSQL("DELETE FROM requested_notifiers")
+            execSQL("DELETE FROM storage_keys")
+            execSQL("DELETE FROM text_primitive_values")
+            execSQL("DELETE FROM types")
+        }
+    }
+
+    fun getOrCreateSchemaTypeId(schema: Schema): TypeId {
+        schemaTypeMap.value[schema.hash]?.let { return it }
+
+        return writableDatabase.useTransaction {
+            val content = ContentValues().apply {
+                put("name", schema.hash)
+                put("is_primitive", false)
+            }
+            val schemaTypeId = insert("types", null, content)
+
+            schemaTypeMap.value[schema.hash] = schemaTypeId
+
+            val insertFieldStatement = compileStatement(
+                "INSERT INTO fields (type_id, parent_type_id, name) VALUES (?, ?, ?)"
+            )
+            val insertFieldBlock = { fieldName: String, fieldType: FieldType ->
+                insertFieldStatement.apply {
+                    bindLong(1, getTypeId(fieldType))
+                    bindLong(2, schemaTypeId)
+                    bindString(3, fieldName)
+                    executeInsert()
+                }
+                Unit
+            }
+            schema.fields.singletons.forEach(insertFieldBlock)
+            schema.fields.collections.forEach(insertFieldBlock)
+            schemaTypeId
+        }
+    }
+
+    /** Returns the type ID for the given [fieldType] if known, otherwise throws. */
+    @VisibleForTesting
+    fun getTypeId(fieldType: FieldType): TypeId = when (fieldType) {
+        is FieldType.Primitive -> fieldType.primitiveType.ordinal.toLong()
+        is FieldType.EntityRef -> schemaTypeMap.value[fieldType.schemaHash]
+            ?: throw IllegalArgumentException(
+                "Unknown type ID for schema with hash ${fieldType.schemaHash}"
+            )
+    }
+
+    /** Loads all schema type IDs from the 'types' table into memory. */
+    private fun loadTypes(): MutableMap<String, TypeId> {
+        val schemaTypeMap = mutableMapOf<String, TypeId>()
+        readableDatabase.rawQuery(
+            "SELECT name, id FROM types WHERE is_primitive = 0",
+            emptyArray()
+        ).use {
+            while (it.moveToNext()) {
+                val hash = it.getString(0)
+                val id = it.getLong(1)
+                schemaTypeMap[hash] = id
+            }
+        }
+        return schemaTypeMap
+    }
+
+    companion object {
+        internal const val DB_NAME = "resurrection.sqlite3"
+        private const val DB_VERSION = 1
+
+        private val CREATE =
+            """
+                CREATE TABLE types (
```

Comment:
It would be helpful to put semicolons after each of these statements, so that it's possible to copy-paste the entire string into a sqlite3 command line client to test things.

Reply:
Good idea!